File: oracle/server/lib/x402.mjs
/**
 * x402 payment gate for GET /api/oracle/parcels/:pin/history
 */
const RECEIVER_ACCOUNT = process.env.HEDERA_ACCOUNT_ID || '0.0.7952768';
const RECEIVER_EVM = (process.env.HEDERA_EVM_ADDRESS || '0x85652f8479dc8dbd89adaee37d42e7c91a534294').toLowerCase();
const PAYMENT_TINYBARS = Number(process.env.X402_PRICE_TINYBARS || '5000000');
const MIRROR_BASE = 'https://testnet.mirrornode.hedera.com';
const MAX_TX_AGE_SECS = 300;
const PROTECTED = [/^\/api\/oracle\/parcels\/[^/]+\/history$/];

const usedTxIds = new Set();

function isProtected(path) {
  return PROTECTED.some((re) => re.test(path));
}

function paymentRequired(resource) {
  return {
    status: 402,
    body: {
      x402Version: 1,
      error: null,
      accepts: [
        {
          scheme: 'hedera-hbar',
          network: 'testnet',
          maxAmountRequired: String(PAYMENT_TINYBARS),
          resource,
          description: 'ZoneProof Oracle — parcel rezoning history',
          mimeType: 'application/json',
          payTo: RECEIVER_ACCOUNT,
          maxTimeoutSeconds: MAX_TX_AGE_SECS,
        },
      ],
    },
    headers: { 'X-402-Version': '1' },
  };
}

function decodePaymentHeader(header) {
  let h = header;
  const pad = 4 - (h.length % 4);
  if (pad !== 4) h += '='.repeat(pad);
  return JSON.parse(Buffer.from(h, 'base64').toString('utf8'));
}

async function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function verifyEvmPayment(txHash) {
  if (usedTxIds.has(txHash)) return [false, 'Transaction already used'];
  const url = `${MIRROR_BASE}/api/v1/contracts/results/${txHash}`;
  let last = 'unknown';
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const r = await fetch(url);
      if (r.status === 404) {
        last = `EVM transaction not yet indexed (attempt ${attempt + 1}/5)`;
        if (attempt < 4) await sleep(2000);
        continue;
      }
      if (!r.ok) return [false, `Mirror node returned HTTP ${r.status}`];
      const data = await r.json();
      if (data.error_message) return [false, `Transaction reverted: ${data.error_message}`];
      const consensusTs = Number.parseFloat(data.timestamp || '0');
      if (Date.now() / 1000 - consensusTs > MAX_TX_AGE_SECS) return [false, 'Payment too old (max 5 minutes)'];
      const toAddr = (data.to || '').toLowerCase();
      if (toAddr !== RECEIVER_EVM) return [false, `Wrong receiver: got ${toAddr}, expected ${RECEIVER_EVM}`];
      const amount = Number(data.amount || 0);
      if (amount < PAYMENT_TINYBARS) return [false, `Insufficient payment: ${amount} tinybars < ${PAYMENT_TINYBARS}`];
      usedTxIds.add(txHash);
      return [true, 'ok'];
    } catch (e) {
      last = `EVM verification error: ${e.message}`;
      if (attempt < 4) await sleep(2000);
    }
  }
  return [false, last];
}

async function verifyHederaPayment(txId) {
  if (usedTxIds.has(txId)) return [false, 'Transaction ID already used'];
  let mirrorTxId = txId;
  if (txId.includes('@')) {
    const [accountPart, tsPart] = txId.split('@', 2);
    mirrorTxId = `${accountPart}-${tsPart.replace('.', '-')}`;
  }
  const url = `${MIRROR_BASE}/api/v1/transactions/${mirrorTxId}`;
  let last = 'unknown';
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const r = await fetch(url);
      if (r.status === 404) {
        last = `Transaction not yet indexed (attempt ${attempt + 1}/5)`;
        if (attempt < 4) await sleep(2000);
        continue;
      }
      if (!r.ok) return [false, `Mirror node returned HTTP ${r.status}`];
      const data = await r.json();
      for (const tx of data.transactions || []) {
        if (tx.result !== 'SUCCESS') continue;
        const consensusTs = Number.parseFloat(tx.consensus_timestamp || '0');
        if (Date.now() / 1000 - consensusTs > MAX_TX_AGE_SECS) return [false, 'Payment too old (max 5 minutes)'];
        for (const transfer of tx.transfers || []) {
          if (transfer.account === RECEIVER_ACCOUNT && (transfer.amount || 0) >= PAYMENT_TINYBARS) {
            usedTxIds.add(txId);
            return [true, 'ok'];
          }
        }
      }
      return [false, 'No qualifying HBAR transfer to app wallet found'];
    } catch (e) {
      last = `Verification failed: ${e.message}`;
      if (attempt < 4) await sleep(2000);
    }
  }
  return [false, last];
}

export function x402Middleware(req, res, next) {
  if (!isProtected(req.path)) return next();

  const xPayment = req.headers['x-payment'];
  if (!xPayment) {
    const pr = paymentRequired(req.path);
    res.set(pr.headers);
    return res.status(pr.status).json(pr.body);
  }

  (async () => {
    try {
      const payment = decodePaymentHeader(xPayment);
      const scheme = payment.scheme || '';
      const txHash = payment.txHash || '';
      const txId = payment.txId || '';
      let ok, reason;
      if (scheme === 'hedera-evm' || String(txHash).startsWith('0x')) {
        [ok, reason] = await verifyEvmPayment(txHash);
      } else {
        [ok, reason] = await verifyHederaPayment(txId);
      }
      if (!ok) {
        return res.status(402).set({ 'X-402-Version': '1' }).json({ x402Version: 1, error: reason });
      }
      next();
    } catch {
      return res.status(402).json({ x402Version: 1, error: 'Malformed X-Payment header' });
    }
  })();
}

File: oracle/api/middleware/x402.py
"""
x402 payment middleware for ZoneProof Oracle API.

Flow:
  1. Client calls a protected endpoint (no X-Payment header)
  2. Middleware returns 402 with Hedera HBAR payment instructions
  3. Client pays HBAR to app wallet, gets TX ID
  4. Client encodes { txId, network } as base64, retries with X-Payment header
  5. Middleware verifies TX on Hedera mirror node → serves response

Protected routes: /api/oracle/parcels/{pin}/history
"""

import re
import json
import base64
import time
import asyncio
import os

import httpx
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse

RECEIVER_ACCOUNT  = os.getenv("HEDERA_ACCOUNT_ID",   "0.0.7952768")
RECEIVER_EVM      = os.getenv("HEDERA_EVM_ADDRESS", "0x85652f8479dc8dbd89adaee37d42e7c91a534294").lower()
PAYMENT_TINYBARS  = int(os.getenv("X402_PRICE_TINYBARS", "5000000"))   # 0.05 HBAR
MIRROR_BASE       = "https://testnet.mirrornode.hedera.com"
MAX_TX_AGE_SECS   = 300   # payment must be within last 5 minutes

# Regex patterns for routes that require payment
PROTECTED_PATTERNS = [
    r"^/api/oracle/parcels/[^/]+/history$",
]

# In-memory replay-attack guard — TX IDs that have already been used
_used_tx_ids: set = set()


def _is_protected(path: str) -> bool:
    return any(re.match(p, path) for p in PROTECTED_PATTERNS)


def _payment_required_response(resource: str) -> JSONResponse:
    """Standard x402 response body."""
    return JSONResponse(
        status_code=402,
        content={
            "x402Version": 1,
            "error": None,
            "accepts": [
                {
                    "scheme":            "hedera-hbar",
                    "network":           "testnet",
                    "maxAmountRequired": str(PAYMENT_TINYBARS),
                    "resource":          resource,
                    "description":       "ZoneProof Oracle — parcel rezoning history",
                    "mimeType":          "application/json",
                    "payTo":             RECEIVER_ACCOUNT,
                    "maxTimeoutSeconds": MAX_TX_AGE_SECS,
                }
            ],
        },
        headers={"X-402-Version": "1"},
    )


def _decode_payment_header(header: str) -> dict:
    # Add base64 padding if needed
    padding = 4 - (len(header) % 4)
    if padding != 4:
        header += "=" * padding
    raw = base64.b64decode(header).decode("utf-8")
    return json.loads(raw)


async def _verify_evm_payment(tx_hash: str) -> tuple:
    """Verify an EVM tx hash (from HashPack via window.ethereum) on the mirror node.

    Retries up to 5 times with 2-second gaps because the Hedera mirror node
    typically lags 3-10 seconds behind on-chain consensus.
    """
    if tx_hash in _used_tx_ids:
        return False, "Transaction already used"

    url = f"{MIRROR_BASE}/api/v1/contracts/results/{tx_hash}"

    last_error = "unknown"
    for attempt in range(5):
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                r = await client.get(url)

            if r.status_code == 404:
                # Mirror node hasn't indexed the tx yet — wait and retry
                last_error = f"EVM transaction not yet indexed (attempt {attempt + 1}/5)"
                if attempt < 4:
                    await asyncio.sleep(2)
                continue

            if r.status_code != 200:
                return False, f"Mirror node returned HTTP {r.status_code}"

            data = r.json()

            # Check success (failed txs should not unlock content)
            error_message = data.get("error_message")
            if error_message:
                return False, f"Transaction reverted: {error_message}"

            # Check timestamp (seconds.nanoseconds string from mirror node)
            consensus_ts = float(data.get("timestamp", "0"))
            if time.time() - consensus_ts > MAX_TX_AGE_SECS:
                return False, "Payment too old (max 5 minutes)"

            # Check receiver matches app EVM address
            to_addr = (data.get("to") or "").lower()
            if to_addr != RECEIVER_EVM:
                return False, f"Wrong receiver: got {to_addr!r}, expected {RECEIVER_EVM!r}"

            # Amount is in tinybars in the mirror node contracts/results response
            # 0.05 HBAR = 5,000,000 tinybars (not wei — Hedera normalises internally)
            amount = int(data.get("amount", 0))
            if amount < PAYMENT_TINYBARS:
                return False, f"Insufficient payment: {amount} tinybars < {PAYMENT_TINYBARS}"

            _used_tx_ids.add(tx_hash)
            return True, "ok"

        except Exception as exc:
            last_error = f"EVM verification error: {exc}"
            if attempt < 4:
                await asyncio.sleep(2)

    return False, last_error


async def _verify_hedera_payment(tx_id: str) -> tuple:
    """Verify a native Hedera HBAR transfer on the testnet mirror node."""
    if tx_id in _used_tx_ids:
        return False, "Transaction ID already used"

    # Mirror node expects: 0.0.XXXXXX-SEC-NANO
    # Client may send:     0.0.XXXXXX@SEC.NANO
    if "@" in tx_id:
        account_part, ts_part = tx_id.split("@", 1)
        ts_part = ts_part.replace(".", "-")
        mirror_tx_id = f"{account_part}-{ts_part}"
    else:
        mirror_tx_id = tx_id

    url = f"{MIRROR_BASE}/api/v1/transactions/{mirror_tx_id}"

    last_error = "unknown"
    for attempt in range(5):
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                r = await client.get(url)

            if r.status_code == 404:
                last_error = f"Transaction not yet indexed (attempt {attempt + 1}/5)"
                if attempt < 4:
                    await asyncio.sleep(2)
                continue

            if r.status_code != 200:
                return False, f"Mirror node returned HTTP {r.status_code}"

            data = r.json()
            transactions = data.get("transactions", [])

            for tx in transactions:
                if tx.get("result") != "SUCCESS":
                    continue

                consensus_ts = float(tx.get("consensus_timestamp", "0"))
                if time.time() - consensus_ts > MAX_TX_AGE_SECS:
                    return False, "Payment too old (max 5 minutes)"

                for transfer in tx.get("transfers", []):
                    if (
                        transfer.get("account") == RECEIVER_ACCOUNT
                        and transfer.get("amount", 0) >= PAYMENT_TINYBARS
                    ):
                        _used_tx_ids.add(tx_id)
                        return True, "ok"

            return False, "No qualifying HBAR transfer to app wallet found"

        except Exception as exc:
            last_error = f"Verification failed: {exc}"
            if attempt < 4:
                await asyncio.sleep(2)

    return False, last_error


class X402Middleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        if not _is_protected(request.url.path):
            return await call_next(request)

        x_payment = request.headers.get("X-Payment")

        if not x_payment:
            return _payment_required_response(request.url.path)

        try:
            payment = _decode_payment_header(x_payment)
            scheme  = payment.get("scheme", "")
            tx_hash = payment.get("txHash", "")  # EVM path (HashPack via window.ethereum)
            tx_id   = payment.get("txId",   "")  # Native Hedera path (MCP server)
        except Exception:
            return JSONResponse(
                status_code=402,
                content={"x402Version": 1, "error": "Malformed X-Payment header"},
            )

        if scheme == "hedera-evm" or tx_hash.startswith("0x"):
            ok, reason = await _verify_evm_payment(tx_hash)
        else:
            ok, reason = await _verify_hedera_payment(tx_id)
        if not ok:
            return JSONResponse(
                status_code=402,
                content={"x402Version": 1, "error": reason},
                headers={"X-402-Version": "1"},
            )

        return await call_next(request)
